### PR TITLE
Support for epoch winning bakers and first block queries.

### PR DESCRIPTION
**Pull request by td202**
## Purpose

Add support for getting the winning bakers of an epoch and the first block of an epoch.

## Changes

- Update `concordium-grpc-api`.
- Add `EpochRequest` and `WinningBaker` types, with `ToProto` support.
- Implement an interpolation search. This is used for finding the first block of an epoch, which is used by both queries.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: concordium-base.cabal [hunk 1]
```diff
@@ -1,6 +1,6 @@
 cabal-version: 2.0
 
--- This file has been generated from package.yaml by hpack version 0.35.0.
+-- This file has been generated from package.yaml by hpack version 0.35.1.
```
Comment (EmilLa1) [inline concordium-base.cabal:3]: Just out of curiosity, is this because of a new stack version we're using or?

**Review by EmilLa1 — APPROVED** [review]
Looking good!

File: haskell-src/Concordium/Utils/InterpolationSearch.hs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    (index -> m (key, value)) ->
+    -- |Key to search for @target@.
+    key ->
+    -- |Lower bound of the search range @(lowIx, (lowKey, lowVal))@.
+    (index, (key, value)) ->
+    -- |Upper bound of the search range @(highIx, (highKey, highVal))@.
+    (index, (key, value)) ->
+    m (Maybe (index, value))
+{-# INLINE interpolationSearchFirstM #-}
+interpolationSearchFirstM lookupIx target low@(lowIx, (lowKey, lowVal)) high@(_, (highKey, _))
```
Comment (abizjak) [inline haskell-src/Concordium/Utils/InterpolationSearch.hs:33]: Could you please add some unit tests for this.

File: haskell-src/Concordium/Types/Queries.hs [hunk 3]
```diff
@@ -1028,3 +1028,30 @@ instance ToJSON BakerRewardPeriodInfo where
...
+          erEpoch :: !Epoch
+        }
+    | -- |Query the epoch of a specified block.
+      EpochOfBlock
+        { -- |The block whose epoch is to be used for the query.
+          erBlock :: !BlockHashInput
+        }
+
+-- |Details of which baker won the lottery in a given round in consensus version 1.
+data WinningBaker = WinningBaker
```
Comment (abizjak) [inline haskell-src/Concordium/Types/Queries.hs:1049]: We should probably add ToJSON instance for this so we can display it in the client.

**Review by abizjak — APPROVED** [review]

File: concordium-base.cabal (hunk 1 above)
Comment (td202) [reply concordium-base.cabal:3]: Yes. `stack` version 2.9.3 uses `hpack` 0.35.1.

File: haskell-src/Concordium/Utils/InterpolationSearch.hs (hunk 2 above)
Comment (td202) [reply haskell-src/Concordium/Utils/InterpolationSearch.hs:35]: Done.

File: haskell-src/Concordium/Types/Queries.hs (hunk 3 above)
Comment (td202) [reply haskell-src/Concordium/Types/Queries.hs:1049]: Done.

**Pull request merged**